File: app/src/main/java/com/billiamram/prioritizer/ItemSetListPresenter.java
package com.billiamram.prioritizer;

import com.billiamram.prioritizer.domain.ItemSet;
import com.billiamram.prioritizer.domain.ItemSetRepository;

import java.util.List;

public class ItemSetListPresenter {
    private View mView;
    private ItemSetRepository itemSetRepository;
    private List<ItemSet> itemSets;

    public ItemSetListPresenter(ItemSetRepository itemSetRepository) {
        this.itemSetRepository = itemSetRepository;
    }

    public void viewLoaded(View view) {
        mView = view;
        itemSets = itemSetRepository.findAll();
        mView.showItemSets(itemSets);
    }

    public void viewUnloaded(View view) {
        if (mView == view) mView = null;
    }

    public void itemSetSelected(int index) {
        mView.navigateToItemSet(itemSets.get(index).id);
    }

    public interface View {
        void showItemSets(List<ItemSet> itemSets);

        void navigateToItemSet(String itemSetId);
    }
}

File: app/src/main/java/com/billiamram/prioritizer/ItemSetListPresenter.kt
package com.billiamram.prioritizer

import com.billiamram.prioritizer.domain.ItemSet
import com.billiamram.prioritizer.domain.ItemSetRepository

class ItemSetListPresenter(private val itemSetRepository: ItemSetRepository) {
    private var mView: View? = null
    private var itemSets: List<ItemSet>? = null

    fun viewLoaded(view: View) {
        mView = view
        itemSets = itemSetRepository.findAll()
        itemSets?.let { mView?.showItemSets(it) }
    }

    fun viewUnloaded(view: View) {
        if (mView === view) mView = null
    }

    fun itemSetSelected(index: Int) {
        itemSets?.let { mView?.navigateToItemSet(it[index].id) }
    }

    interface View {
        fun showItemSets(itemSets: List<ItemSet>)

        fun navigateToItemSet(itemSetId: String)
    }
}
